third_party/move/move-compiler-v2/src/pipeline/explicit_drop.rs

```diff
@@ -0,0 +1,197 @@
+};
+use move_binary_format::file_format::CodeOffset;
+use move_model::{ast::TempIndex, model::FunctionEnv, ty::Type};
+use move_stackless_bytecode::{
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{AssignKind, AttrId, Bytecode, Operation},
+};
+use std::collections::BTreeSet;
+
+pub struct ExplicitDrop {}
+
+impl FunctionTargetProcessor for ExplicitDrop {
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        fun_env: &FunctionEnv,
+        mut data: FunctionData,
+        _scc_opt: Option<&[FunctionEnv]>,
+    ) -> FunctionData {
+        if fun_env.is_native() {
+            return data;
+        }
+        let target = FunctionTarget::new(fun_env, &data);
+        let mut transformer = ExplicitDropTransformer::new(target);
+        transformer.transform();
+        data.code = transformer.transformed;
+        data.annotations.remove::<LiveVarAnnotation>();
+        data.annotations.remove::<LifetimeAnnotation>();
+        data
+    }
+
+    fn name(&self) -> String {
+        "ExplicitDrop".to_owned()
+    }
+}
+
+struct ExplicitDropTransformer<'a> {
+    target: FunctionTarget<'a>,
+    // result of the transformation
+    transformed: Vec<Bytecode>,
+    live_var_annot: &'a LiveVarAnnotation,
+    lifetime_annot: &'a LifetimeAnnotation,
+}
+
+impl<'a> ExplicitDropTransformer<'a> {
+    pub fn new(target: FunctionTarget<'a>) -> Self {
+        let live_var_annot = target
+            .get_annotations()
+            .get::<LiveVarAnnotation>()
+            .expect("livevar annotation");
+        let lifetime_annot = target
+            .get_annotations()
+            .get::<LifetimeAnnotation>()
+            .expect("lifetime annotation");
+        ExplicitDropTransformer {
+            target,
+            transformed: Vec::new(),
+            live_var_annot,
+            lifetime_annot,
+        }
+    }
+
+    /// Add explicit drop instructions
+    /// note that this will invalidate existing analyses
+    pub fn transform(&mut self) {
+        self.drop_unused_args();
+        for (code_offset, bytecode) in self.target.get_bytecode().to_vec().iter().enumerate() {
+            self.emit_bytecode(bytecode.clone());
+            self.explicate_drops_at(code_offset as CodeOffset, bytecode);
+        }
+    }
+
+    // add drops at given code offset
+    fn explicate_drops_at(&mut self, code_offset: CodeOffset, bytecode: &Bytecode) {
+        match bytecode {
+            Bytecode::Ret(..) | Bytecode::Jump(..) | Bytecode::Abort(..) | Bytecode::Branch(..) => {
+                ()
+            },
+            _ => {
+                let released_temps = self.released_temps_at(code_offset);
+                self.drop_temps(&released_temps, bytecode.get_attr_id())
+            },
+        }
+    }
+
+    fn drop_unused_args(&mut self) {
```

Comment:
Please take a quick look at undocumented methods and consider adding docs.

Reply:
done